backend/lib/kubernetes-client/cache/Reflector.js

```diff
@@ -0,0 +1,339 @@
+    // situations we may force listing directly from etcd (by setting ResourceVersion="")
+    // which will return paginated result, even if watch cache is enabled. However, in
+    // that case, we still want to prefer sending requests to watch cache if possible.
+    //
+    // Paginated result returned for request with ResourceVersion="0" mean that watch
+    // cache is disabled and there are a lot of objects of a given type. In such case,
+    // there is no need to prefer listing from watch cache.
+    if (options.resourceVersion === '0' && paginatedResult) {
+      this.paginatedResult = true
+    }
+
+    this.isLastSyncResourceVersionExpired = false
+    this.store.replace(list.items)
+    this.lastSyncResourceVersion = resourceVersion
+    try {
+      while (!this.stopRequested) {
+        const options = {
+          allowWatchBookmarks: true,
+          timeoutSeconds: randomize(this.minWatchTimeout.asSeconds()),
+          resourceVersion: this.lastSyncResourceVersion
+        }
+        try {
+          this.socket = this.listWatcher.watch(options)
+          await pEvent(this.socket, 'open', { timeout: this.connectTimeout.asMilliseconds() })
+        } catch (err) {
+          if (isExpiredError(err)) {
+            // Don't set LastSyncResourceVersionExpired - LIST call with ResourceVersion=RV already
+            // has a semantic that it returns data at least as fresh as provided RV.
+            // So first try to LIST with setting RV to resource version of last observed object.
+            logger.info('%s: watch of %s not opened with: %s', this.name, this.expectedTypeName, err.message)
+          } else {
+            logger.error('%s: watch of %s failed with: %s', this.name, this.expectedTypeName, err)
+          }
+          if (includes(RETRY_ERROR_CODES, err.code)) {
+            await delay(randomize(this.period.asMilliseconds()))
+            continue
+          }
+          return
+        }
+        if (this.stopRequested) {
+          return
+        }
+        // run websocket heartbeat
+        (async () => {
+          try {
+            await this.heartbeat(this.socket)
+          } catch (err) { /* ignore error */ }
+        })()
+        // handle websocket messages
+        try {
+          await this.watchHandler(this.socket)
+        } catch (err) {
+          if (isExpiredError(err)) {
+            // Don't set LastSyncResourceVersionExpired - LIST call with ResourceVersion=RV already
+            // has a semantic that it returns data at least as fresh as provided RV.
+            // So first try to LIST with setting RV to resource version of last observed object.
+            logger.info('%s: watch of %s closed with: %s', this.name, this.expectedTypeName, err.message)
+          } else {
+            logger.warn('%s: watch of %s ended with: %s', this.name, this.expectedTypeName, err)
+          }
+          return
+        }
```

Comment:
throw expection / assert that socket is closed

Reply:
I decided to finally do a graceful socket close at the end of every watch iteration. Therefore we can be sure the socket is closed before we enter the next retry or reWatch or reListAndWatch.